prevayler4/git-coherence/src/house/jux__/prevayler4/git_coherence__.clj

```diff
@@ -0,0 +1,126 @@
+    (doseq [candidate candidates]
+      (remove-ns (ns-name candidate)))))
+
+(defn- load! [required-commit opts]
+  (git-restore required-commit opts)
+  (binding [*ns* (find-ns 'house.jux--.prevayler4.git-coherence--)] ; Any ns just to satisfy refresh's expectation of running in the repl.
+    (repl/refresh))
+  (unload-deleted-namespaces opts))
+
+(defn commit-change-event [old-hash]
+  (let [new-hash (git "rev-parse" "HEAD")]
+    (when-not (= new-hash old-hash)
+      {:current-commit-hash new-hash})))
+
+(defn- ignoring-coherence [business-fn]
+  (fn [state event timestamp]
+    (if (:current-commit-hash event)
+      state
+      (business-fn state event timestamp))))
+
+(defn- providing-coherence [business-fn opts]
+  (let [loaded-commit (atom nil)]
+    (fn [state event timestamp]
+      (let [new-commit (:current-commit-hash event) ; Will be nil when event is a regular business event.
+            state (if new-commit
+                    (assoc state :current-commit-hash new-commit)
+                    state)
+            required-commit (:current-commit-hash state)]
+        (check required-commit "Journal event must be replayed with an associated commit.")
+        (when-not (= required-commit @loaded-commit) ; Works for the first event also (current-commit-atom is nil).
+          (load! required-commit opts)
+          (reset! loaded-commit required-commit))
+        (if new-commit
+          state
+          (business-fn state event timestamp))))))
+
+(defn- wrap-for-code-version-coherence [business-fn {:keys [coherent-mode?] :as opts}]
+  (if coherent-mode?
+    (providing-coherence business-fn opts)
+    (ignoring-coherence  business-fn)))
+
+(defn- workspace-dirty? []
+  (try
+    (run "git" "diff" "--quiet")
+    false
+    (catch Exception e
+      (if (-> e ex-data :result :exit (= 1))
+        true
+        (throw e)))))
+
+(defn start!
+  "Starts a prevayler instance that has git coherence enabled.
+
+   Arguments:
+   - start-prevayler-fn: a function that starts the underlying prevayler instance,
+                         it must receive one argument containing the prevayler config
+   - config: the prevayler config to be passed to the underlying prevayler instance
+   - opts: a map with the following options
+     - coherent-mode?: whether coherence should be enabled or not
+     - git-reset?: whether the git workspace should be reset
+     - src-dir: a java.io.File that refers to the directory that points to the source directory,
+                currently only one level directory is supported 
+                (e.g. \"src\" is ok, but \"src/clj\" is not
+     - refreshable-namespaces: a sequence of namespace symbols that need to be refreshed,
+                               all namespaces under the given symbol will be refreshed
+                               (e.g. ['my-system.biz] will refresh 'my-system.biz, 
+                               'my-system.biz.somenamespace and so on)"
```

Comment:
please, have a thorough look at the api of this function (option names, and so on)

Reply:
Sure thing